### PR TITLE
Alternative message reporting modes

### PR DESCRIPTION
**Pull request by kLabz**
Adds alternative message reporting modes with `-D message-reporting=[mode]`

## Reporting mode: "classic"

Current output, still default reporting mode

## Reporting mode: "pretty"

Adds more contextual information to errors (sources) for example like that:

image

Additionally, `-D no-color` can be used if your terminal doesn't support ANSI escape codes:
```
[ERROR] Main.hx:8: characters 9-18

 8 |         C.f("hi");
   |         ^^^^^^^^^
   | Could not find a suitable overload, reasons follow

      | Overload resolution failed for () -> Void

       8 |         C.f("hi");
         |             ^^^^
         | Too many arguments

      | Overload resolution failed for (t : f.T) -> Void

       8 |         C.f("hi");
         |             ^^^^
         | Constraint check failure for f.T
         | String should be Int
         | For function argument 't'
```
_(from actual [test output for 6065](https://github.com/HaxeFoundation/haxe/pull/10863/files#diff-7c25bc4deac5eedeea4b3f30cf9ddf7d472728b37a1a1c3d988fd3b655e06c01))_

## Reporting mode: "indent"

New mode that uses nesting level data in messages to provide more context to tools exploiting compiler output (vscode tasks, for example). This format is experimental and subject to many changes until the whole feature is done on both sides.

Example:
```
$ haxe compile-fail.hxml -D message-reporting=indent
Main.hx:8: characters 3-12 : Could not find a suitable overload, reasons follow
  Main.hx:8: characters 3-12 : Overload resolution failed for () -> Void
    Main.hx:8: characters 7-11 : Too many arguments
  Main.hx:8: characters 3-12 : Overload resolution failed for (t : f.T) -> Void
    Main.hx:8: characters 7-11 : Constraint check failure for f.T
      Main.hx:8: characters 7-11 : String should be Int
      Main.hx:8: characters 7-11 : For function argument 't'
```

## Additional logging to file

Also adds `-D messages-log-file=[path]` (and `-D messages-log-format=[format]`, default being "indent") to make compiler output to a file _in addition_ to terminal output, so that for example you can get user-readable output in terminal and computer-readable output in the log file (to be consumed by tools).

## Next steps after merge

* Update vshaxe to make use of new reporting modes
* Handle interp errors in pretty mode (displayed as "classic" for now)
* Merge successive identical (message + pos) messages, for example `Called from here (x10)`

## Review

I guess `Files changed 110` / `+1,085 -317` won't exactly help getting review :sweat_smile: 

Main changes:
 - `server.ml` holds the actual message reporting modes code
 - compiler errors have been modified with added (optional, defaults to `0` which is top level) `depth` 
 - compiler errors are now handled as `string * pos` at most places instead of `string` with separate `pos`. This was mostly necessary for stacks
 - misc tests have been updated for rare changes and new ones have been added for new outputs; can check those to get an idea what this PR does

---

Grab nightlies to try it out: [`58ee66d` nightlies](https://github.com/HaxeFoundation/haxe/actions/runs/4416017693#artifacts)

### CODE REVIEW

**Comment by Aurel300** [comment]
I like this. Small nitpick: I think this feature is about pretty-printing errors, not about verbose errors. From the latter I would expect more error information, which is not the case here, it is the same information presented in a more readable way. So maybe the define could be `pretty-errors` or similar?

**Comment by kLabz** [comment]
Good point. Maybe it could become both to some extent but I'm not sure it's something we want there nor if it would be practical..

**Comment by Aurel300** [comment]
By the way, this might be out of scope here, but: I find the "hint" which says `for function argument x` really useless most of the time, because it does not give any indication *which* function has the argument `x`. This is made worse by the fact that argument names in the standard library are generally not descriptive. If this kind of hint (in "pretty" mode) could point to the declaration of the function or closure variable, that would be great!

**Comment by kLabz** [comment]
Yeah, trying to do something about it; should be possible if I find a nice way to handle Stack errors. There are similar cases where I want to make some improvements too, like https://github.com/HaxeFoundation/haxe/blob/development/src/typing/typeloadCheck.ml#L503 which is just repeating the function name it's pointing to (redundant with pretty errors)

File: src-json/define.json [hunk 1]
```diff
@@ -718,5 +718,27 @@
...
+	},
+	{
+		"name": "NoColor",
+		"define": "no-color",
+		"doc": "Disable ANSI color codes when using rich output."
+	},
+	{
+		"name": "MessagesLogFile",
+		"define": "messages-log-file",
+		"doc": "Path to a text file to write messages log to."
```
Comment (Aurel300) [inline src-json/define.json]: Should have the same note that you have in the PR: this is *in addition* to the regular output.

File: src/codegen/gencommon/gencommon.ml [hunk 2]
```diff
@@ -732,7 +732,7 @@ let run_filters_from gen t filters =
 let run_filters gen =
 	let last_error = gen.gcon.error in
 	let has_errors = ref false in
-	gen.gcon.error <- (fun msg pos -> has_errors := true; last_error msg pos);
+	gen.gcon.error <- (fun ?depth msg pos -> has_errors := true; last_error msg pos);
```
Comment (Aurel300) [inline src/codegen/gencommon/gencommon.ml]: Should the `depth` arg be passed to `last_error`?

File: tests/misc/projects/Issue8303/pretty-fail.hxml.stderr [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+       | Called from here
+
+       | Called from here
+
+       | Called from here
+
+       | Called from here
+
+       | Called from here
```
Comment (Aurel300) [inline tests/misc/projects/Issue8303/pretty-fail.hxml.stderr:45]: Are these here because the position is unknown, or because it is the same as the previous one? This is an edge case where the pretty reporting only makes things worse… I wonder if the positions should be repeated though.

Comment (kLabz) [reply tests/misc/projects/Issue8303/pretty-fail.hxml.stderr:45]: Same as previous pos. I wanted to merge such identical msg+pos into something like `msg (x10)` but it's kinda low priority

**Comment by kLabz** [comment]
@Simn if you happen to be bored, I think this PR is ready for review

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -347,6 +347,7 @@ jobs:
       PLATFORM: linux64
       TEST: ${{matrix.target}}
       HXCPP_COMPILE_CACHE: ~/hxcache
+      HAXE_STD_PATH: /usr/local/share/haxe/std
```
Comment (Simn) [inline .github/workflows/main.yml:350]: Why is this necessary?

File: src/compiler/compiler.ml [hunk 5]
```diff
@@ -221,17 +221,17 @@ module Setup = struct
...
 					Printf.sprintf "(%s) %s" wobj.w_name msg
 				in
-				message ctx (msg,p,DKCompilerMessage,Warning)
+				message ctx (msg,p,depth,DKCompilerMessage,Warning)
 			| WMDisable ->
 				()
 		);
 		com.error <- error ctx;
-		let filter_messages = (fun keep_errors predicate -> (List.filter (fun ((_,_,_,sev) as cm) ->
+		let filter_messages = (fun keep_errors predicate -> (List.filter (fun ((_,_,_,_,sev) as cm) ->
```
Comment (Simn) [inline src/compiler/compiler.ml]: As a future TODO, I really prefer records over these terror tuples. My personal cutoff is more than three items.

File: src/context/typecore.ml [hunk 6]
```diff
@@ -197,7 +197,7 @@ type dot_path_part = {
 
 exception Forbid_package of (string * path * pos) * pos list * string
 
-exception WithTypeError of error_msg * pos
+exception WithTypeError of error_msg * pos * int
```
Comment (Simn) [inline src/context/typecore.ml]: A comment to note what this `int` means would be good here.

File: src/core/error.ml [hunk 7]
```diff
@@ -43,6 +43,9 @@ let short_type ctx t =
 	Should be called for each complementary error message.
 *)
 let compl_msg s = "... " ^ s
+let rec compl_located_msg = function
```
Comment (Simn) [inline src/core/error.ml:46]: wtf is "compl"? I know it's there already, but that's a strange abbreviation.

File: src/core/error.ml [hunk 8]
```diff
@@ -17,7 +17,7 @@ and error_msg =
 	| Unify of unify_error list
 	| Custom of string
 	| Unknown_ident of string
-	| Stack of error_msg * error_msg
+	| Stack of (error_msg * Globals.pos) list
```
Comment (Simn) [inline src/core/error.ml:20]: Is this really easier to handle? The problem is that you can still have `Stack [Stack []]` kind of things, so you always have to manage the recursive case anyway.

File: src/typing/typer.ml [hunk 9]
```diff
@@ -508,9 +508,9 @@ and type_ident ctx i p mode with_type =
 						AKExpr (mk (TIdent i) t p)
 					end else match ctx.com.display.dms_kind with
 						| DMNone ->
-							raise (Error(err,p))
+							raise (Error(err,p,0))
 						| _ ->
-							display_error ctx.com (error_msg err) p;
+								located_display_error ctx.com (error_msg p err);
```
Comment (Simn) [inline src/typing/typer.ml]: indentation

**Comment by Simn** [comment]
I like the `depth` approach, that seems quite pragmatic and easy to handle.

I kinda wish we didn't do this stupid tuple stuff so much because it's so awkward to extend. It feels like half of the changes are additional `,_` which is quite silly. I do understand if you don't want to address this as part of the PR though.

The diff here would also benefit quite a bit from a separate "whitespace cleanup" PR, but it's fine if you can't be arsed to factor that out.

File: .github/workflows/main.yml (hunk 4 above)
Comment (kLabz) [reply .github/workflows/main.yml:350]: On unix, when this is not set we end up with 3 std path https://github.com/HaxeFoundation/haxe/blob/development/src/compiler/compiler.ml#L185

When our CI runs misc tests on mac, the second will be used for eval error reporting; on linux, the third one (and on windows, the first and only one).

Setting `HAXE_STD_PATH` allowed me to have a predictable std path to provide to misc tests and use to check error reporting [on std types](https://github.com/HaxeFoundation/haxe/blob/ca40974ad12828d2d1caccee9cd2c79e74dc156e/tests/misc/projects/Issue10844/user-defined-meta-json-indent-fail.hxml.stderr)

... which reminds me I couldn't add expected test output for those back in the day; will add that now.
Do you want me to extract this std path exposed to misc test business in a separate PR?

File: src/core/error.ml (hunk 7 above)
Comment (kLabz) [reply src/core/error.ml:46]: Complementary information or something like that? If you have a better idea I can update those on this PR

File: src/core/error.ml (hunk 8 above)
Comment (kLabz) [reply src/core/error.ml:20]: Previously this was handled with ?inner positions" printed in sub errors and stored as plain string. Now I want to keep error message and positions separate to be able to report those any way I want at the end, so I need both position and message for each stack entry.
Are you suggesting I use `Stack of (error_msg * pos) * (error_msg * pos)` and nest Stack items in there?

Comment (Simn) [reply src/core/error.ml:20]: Well, that would be the obvious change. But if you find the `list` approach easier to handle then I don't mind.

Comment (kLabz) [reply src/core/error.ml:20]: Will check if that's annoying to use or not.

Comment (kLabz) [reply src/core/error.ml:20]: Yeah I do find the list approach easier to handle, especially when dealing with conversions
